app/lib/frontend/dom/dom.dart

```diff
@@ -0,0 +1,160 @@
+DomContext dom = StringDomContext();
+
+/// Opaque entity for DOM nodes.
+abstract class Node {}
+
+/// Factory class to create DOM [Node]s.
+abstract class DomContext {
+  /// Creates a DOM fragment from the list of [children] nodes.
+  Node fragment(List<Node> children);
+
+  /// Creates a DOM Element.
+  Node element(
+    String tag, {
+    String id,
+    List<String> classes,
+    Map<String, String> attributes,
+    List<Node> children,
+  });
+
+  /// Creates a DOM Text node.
+  Node text(String value);
+}
+
+extension DomContextExt on DomContext {
+  Node div({
+    String id,
+    List<String> classes,
+    Map<String, String> attributes,
+    List<Node> children,
+  }) =>
+      element(
+        'div',
+        id: id,
+        classes: classes,
+        attributes: attributes,
+        children: children,
+      );
+
+  Node span({
+    String id,
+    List<String> classes,
+    Map<String, String> attributes,
+    List<Node> children,
+  }) =>
+      element(
+        'span',
+        id: id,
+        classes: classes,
+        attributes: attributes,
+        children: children,
+      );
+}
+
+/// Uses DOM nodes to emit escaped HTML string.
+class StringDomContext extends DomContext {
+  @override
+  Node fragment(List<Node> children) => _StringNodeList(children);
+
+  @override
+  Node element(
+    String tag, {
+    String id,
+    List<String> classes,
+    Map<String, String> attributes,
+    List<Node> children,
+  }) =>
+      _StringElement(tag, _mergeAttributes(id, classes, attributes), children);
+
+  @override
+  Node text(String value) => _StringText(value);
+}
+
+Map<String, String> _mergeAttributes(
+    String id, List<String> classes, Map<String, String> attributes) {
+  final hasClasses = classes != null && classes.isNotEmpty;
+  final hasAttributes =
+      id != null || hasClasses || (attributes != null && attributes.isNotEmpty);
+  if (!hasAttributes) return null;
+  return <String, String>{
+    if (id != null) 'id': id,
+    if (hasClasses) 'class': classes.join(' '),
+    if (attributes != null) ...attributes,
+  };
+}
+
+abstract class _StringNode extends Node {
+  void writeHtml(StringSink sink);
+
+  @override
+  String toString() {
+    final sb = StringBuffer();
+    writeHtml(sb);
+    return sb.toString();
+  }
+}
+
+class _StringNodeList extends _StringNode {
+  final List<Node> children;
+
+  _StringNodeList(this.children);
+
+  @override
+  void writeHtml(StringSink sink) {
+    for (final node in children) {
+      (node as _StringNode).writeHtml(sink);
+    }
+  }
+}
+
+class _StringElement extends _StringNode {
+  final String tag;
+  final Map<String, String> attributes;
+  final List<Node> children;
+
+  _StringElement(this.tag, this.attributes, this.children);
+
+  @override
+  void writeHtml(StringSink sink) {
+    sink.write('<$tag');
+    if (attributes != null) {
+      for (final e in attributes.entries) {
+        sink.write(' ${e.key}="${_attributeEscape.convert(e.value)}"');
```

Comment:
This is probably a good place to point out that it's important to escape `e.key` too ? :D

Or have something in the constructor that throws an `ArgumentError` if `attributes` contains an illegal key.